File: android/app/src/main/java/gh/group38/smartsocket/data/SocketStatus.kt
package gh.group38.smartsocket.data

/**
 * Mirrors `SocketState` in `SmartSocket/src/core/Types.h`.
 *
 * The ordinals are the wire format - the firmware sends the number, not the
 * name, because a 32 KB part has better uses for its flash than strings the
 * phone already has. **If the enum in Types.h gains a member anywhere but the
 * end, this list has to move with it.**
 */
enum class SocketState(val label: String, val detail: String) {
    CALIBRATING("Calibrating", "Measuring the sensor's zero point"),
    READY("Ready", "Waiting for something to be plugged in"),
    SETTLING("Settling", "Ignoring the charger's inrush"),
    CHARGING("Charging", "Watching for the current to taper"),
    CUTOFF("Power cut", "Charged. Power is off"),
    FAULT("Overcurrent", "Something is wrong. Unplug at the wall"),
    MANUAL_OFF("Switched off", "You turned it off"),
    PROBING("Checking", "Looking for a load"),
    RELAY_STUCK("Relay stuck", "Power is still flowing. Unplug at the wall");

    val isPowerOn: Boolean
        get() = this == READY || this == SETTLING || this == CHARGING || this == PROBING

    val isAlarm: Boolean
        get() = this == FAULT || this == RELAY_STUCK

    companion object {
        fun fromOrdinal(value: Int): SocketState? = entries.getOrNull(value)
    }
}

/** One decoded status line. */
data class SocketStatus(
    val state: SocketState,
    val currentMa: Int,
    val peakMa: Int,
    val thresholdMa: Int,
    val sessionElapsedMs: Long,
    val cutoffCount: Int,
    val totalSavedMs: Long,
    val relayClosed: Boolean,
) {
    val amps: Float get() = currentMa / 1000f
    val peakAmps: Float get() = peakMa / 1000f
    val thresholdAmps: Float get() = thresholdMa / 1000f

    companion object {
        val UNKNOWN = SocketStatus(
            state = SocketState.CALIBRATING,
            currentMa = 0,
            peakMa = 0,
            thresholdMa = 0,
            sessionElapsedMs = 0,
            cutoffCount = 0,
            totalSavedMs = 0,
            relayClosed = false,
        )
    }
}

/** Commands the socket accepts. One byte each - see `HalTelemetry::pump`. */
enum class SocketCommand(val wire: String) {
    CUT("C"),
    REARM("R"),
    PROBE("P"),
    STATUS_NOW("?"),
}

File: desktop/src/SmartSocket.Core/SocketStatus.cs
namespace SmartSocket.Core;

/// <summary>
/// Mirrors <c>SocketState</c> in <c>SmartSocket/src/core/Types.h</c>, and
/// <c>SocketState</c> in the Android app.
///
/// The ordinals are the wire format - the firmware sends the number, not the
/// name, because a 32 KB part has better uses for its flash than strings the
/// client already has. <b>If the enum in Types.h gains a member anywhere but
/// the end, this list has to move with it.</b>
/// </summary>
public enum SocketState
{
    Calibrating = 0,
    Ready = 1,
    Settling = 2,
    Charging = 3,
    Cutoff = 4,
    Fault = 5,
    ManualOff = 6,
    Probing = 7,
    RelayStuck = 8,
}

public static class SocketStateExtensions
{
    /// <summary>Highest ordinal the firmware can send.</summary>
    public const int MaxOrdinal = (int)SocketState.RelayStuck;

    public static string Label(this SocketState state) => state switch
    {
        SocketState.Calibrating => "Calibrating",
        SocketState.Ready => "Ready",
        SocketState.Settling => "Settling",
        SocketState.Charging => "Charging",
        SocketState.Cutoff => "Power cut",
        SocketState.Fault => "Overcurrent",
        SocketState.ManualOff => "Switched off",
        SocketState.Probing => "Checking",
        SocketState.RelayStuck => "Relay stuck",
        _ => "Unknown",
    };

    public static string Detail(this SocketState state) => state switch
    {
        SocketState.Calibrating => "Measuring the sensor's zero point",
        SocketState.Ready => "Waiting for something to be plugged in",
        SocketState.Settling => "Ignoring the charger's inrush",
        SocketState.Charging => "Watching for the current to taper",
        SocketState.Cutoff => "Charged. Power is off",
        SocketState.Fault => "Something is wrong. Unplug at the wall",
        SocketState.ManualOff => "You turned it off",
        SocketState.Probing => "Looking for a load",
        SocketState.RelayStuck => "Power is still flowing. Unplug at the wall",
        _ => string.Empty,
    };

    public static bool IsPowerOn(this SocketState state) =>
        state is SocketState.Ready
            or SocketState.Settling
            or SocketState.Charging
            or SocketState.Probing;

    public static bool IsAlarm(this SocketState state) =>
        state is SocketState.Fault or SocketState.RelayStuck;

    /// <summary>
    /// Null for an ordinal this build does not know. An unknown ordinal means
    /// the firmware and this app have drifted apart, and inventing a state for
    /// it would hide exactly that.
    /// </summary>
    public static SocketState? FromOrdinal(int value) =>
        value >= 0 && value <= MaxOrdinal ? (SocketState)value : null;
}

/// <summary>One decoded status line.</summary>
public sealed record SocketStatus(
    SocketState State,
    int CurrentMa,
    int PeakMa,
    int ThresholdMa,
    long SessionElapsedMs,
    int CutoffCount,
    long TotalSavedMs,
    bool RelayClosed)
{
    public float Amps => CurrentMa / 1000f;
    public float PeakAmps => PeakMa / 1000f;
    public float ThresholdAmps => ThresholdMa / 1000f;

    public static readonly SocketStatus Unknown = new(
        State: SocketState.Calibrating,
        CurrentMa: 0,
        PeakMa: 0,
        ThresholdMa: 0,
        SessionElapsedMs: 0,
        CutoffCount: 0,
        TotalSavedMs: 0,
        RelayClosed: false);
}

/// <summary>
/// Commands the socket accepts. One byte each - see <c>HalTelemetry::pump</c>,
/// which only acts on a line that is exactly one character long.
/// </summary>
public enum SocketCommand
{
    Cut,
    Rearm,
    Probe,
    StatusNow,

    /// <summary>
    /// Zero the socket's lifetime cutoff count and saved-time total. They live
    /// in the socket's EEPROM, not in the app, so clearing the app's history
    /// left them untouched and looking stuck.
    /// </summary>
    ResetTotals,
}

public static class SocketCommandExtensions
{
    public static string Wire(this SocketCommand command) => command switch
    {
        SocketCommand.Cut => "C",
        SocketCommand.Rearm => "R",
        SocketCommand.Probe => "P",
        SocketCommand.StatusNow => "?",
        SocketCommand.ResetTotals => "Z",
        _ => throw new ArgumentOutOfRangeException(nameof(command)),
    };
}
